### PR TITLE
Add and verify consensus config hash

### LINKED ISSUE #950: Sanity check version numbers and configs when connecting
Peers need to be running same software version with compatible configs.

When establishing a new connection, we should do a handshake which checks that:
- We're using same version of software
- Hashes of consensus-relevant parts of server configs (e.g. denominations, threshold pubkey) are the same. This would exclude things like bind addresses which would be different for different peers.
- Crash if any differences detected

### PR DESCRIPTION
**Pull request by jkitman**
Makes consensus and client configs implement `Encodable` so they can be deterministically hashed and validated.

Adds the consensus hash and an API endpoint for retrieving it with the client config (next step will be to save and return the signed hash of the client config).

Fixes #950, will be used to solve #1262

### CODE REVIEW

File: fedimint-server/src/lib.rs
Comment (elsirion) [inline fedimint-server/src/lib.rs:115]: Isn't there a query strategy for this?

**Review by elsirion — APPROVED** [review]
Second review welcome since I'm a bit exhausted.

File: fedimint-server/src/lib.rs
Comment (jkitman) [reply fedimint-server/src/lib.rs:115]: The problem is when the federation first starts they all "agree" on the endpoint being down.

I could try to modify the Retry404 to be more generic in ignoring errors.

File: client/client-lib/src/wallet/mod.rs
Comment (dpc) [inline client/client-lib/src/wallet/mod.rs:323]: nit: Should be `unimplement!()`  . AFAIR, `todo` has a meaning - "I was in rush and I'm planning implement it later", while `unimplement` is more "not supported and might never be".

File: fedimint-api/src/encoding/tbs.rs
Comment (dpc) [inline fedimint-api/src/encoding/tbs.rs:42]: Using `SerdeEncodable` in consensus is to be avoided, because it pulls in bincode we we don't generally trust as much. Giving at a blanket impl could invite accidental use.

Possibly you should use `SerdeEncodable` directly on the field you need to encode, and we should think if it doesn't affect consensus.

Comment (dpc) [reply fedimint-api/src/encoding/tbs.rs:42]: Ping @elsirion for second opinion.

It might be a good reason to impl a to/from bytes function for `threshold_crypto` types directly, and then use them here, instead of `SerdeEncodable`.

File: fedimint-server/src/config.rs
Comment (dpc) [inline fedimint-server/src/config.rs:170]: Ad-hoc hashing by listing each field seem very error prone. We should either hash the whole thing, or if not possible for some reason - have a method that all parts of the code can re-use.

File: modules/fedimint-dummy/src/lib.rs
Comment (dpc) [inline modules/fedimint-dummy/src/lib.rs:131]: It's not this PR specific, but we should require users to gtouch `serde_json::Value` in the implementations of these traits. We should have associated types for everything. Will create an issue.

Comment (dpc) [reply modules/fedimint-dummy/src/lib.rs:131]: #1297

Generally when working with it I often get lost in how many types (split configs, aggregates configs) of the `serde_json::Value` we have, all named in rather similar way. Something to think about - maybe there's a way to improve it.

File: modules/fedimint-ln/src/config.rs
Comment (dpc) [inline modules/fedimint-ln/src/config.rs:32]: That `+1` always gets me. It starts with https://docs.rs/hbbft/latest/hbbft/ , where they already use `+ 1` all over the place. They should have made the `threshold` inclusive, because that would be more intuitive.

The fact that we have some "threshold"s mean "other thresholds + 1" is kind of crazy.

We might want to name them `threshold_inclusive" or something. Otherwise it doesn't make sense.

File: fedimint-server/src/lib.rs
Comment (elsirion) [reply fedimint-server/src/lib.rs:115]: "We are all down, so we stay all down!" :laughing:

File: fedimint-api/src/encoding/tbs.rs
Comment (elsirion) [reply fedimint-api/src/encoding/tbs.rs:42]: Kinda agree, can we just impl a `to_bytes` fn in our fork of `threshold_crypto`? Might be easiest.

File: modules/fedimint-ln/src/config.rs
Comment (elsirion) [reply modules/fedimint-ln/src/config.rs:32]: It's just as crazy in the academic literature. I once had to change the definition of `t` in a paper because my prof liked one version over the other which I had found in another paper :man_facepalming:

File: fedimint-api/src/encoding/tbs.rs
Comment (dpc) [reply fedimint-api/src/encoding/tbs.rs:42]: Let's mimic the naming that secp256k1 use: https://docs.rs/secp256k1/latest/secp256k1/struct.SecretKey.html#method.secret_bytes @jkitman 

(sorry for the detour! :D )

Comment (jkitman) [reply fedimint-api/src/encoding/tbs.rs:42]: I will expose a method to get the coefficients then encode them, similar to what we do in the TBS crate.

File: fedimint-server/src/config.rs
Comment (jkitman) [reply fedimint-server/src/config.rs:170]: I agree, but I would need to modify the `Encodable` macro so we could ignore certain fields like the untyped module json and encode it using the typed version instead.  I'm not sure how to do that but maybe you could?

Comment (dpc) [reply fedimint-server/src/config.rs:174]: We can implement `Encodable` manually if need be. Not sure if it helps here, but I think it might.

Comment (dpc) [reply fedimint-server/src/config.rs:174]: Alternatively, we should convert the config to another version that is purely encodable before hashing. We only need to be able to do it one way anyway, I think?

We should totally `impl (En)Decodable for serde_json::Value`, BTW!

**Comment by elsirion** [comment]
triggered CI

File: fedimint-server/src/config.rs
Comment (elsirion) [reply fedimint-server/src/config.rs:174]: > We should totally `impl (En)Decodable for serde_json::Value`, BTW!

Maybe sort term, long term I'd much rather make the type erased cfg a dyn newtype that can be en/decoded and serialized/deserialized. That would require another decoder fn, but that's ok imo.

**Comment by elsirion** [comment]
Needs a rebase after #1323 :frowning:

File: client/client-lib/src/api.rs
Comment (dpc) [inline client/client-lib/src/api.rs:427]: This should be available on `DynFederationApi` after that big refactor landed.

File: fedimint-api/src/encoding/btc.rs
Comment (dpc) [inline fedimint-api/src/encoding/btc.rs:43]: How stable and future-proof is the serde representation of miniscrit's Descriptor? Could we use maybe something more stable?

Possibly maybe we should implement `Encodable` for this ourselves, explicitly handling each variant we support so we can make sure constants don't change?

I don't know ... just thinking aloud. In 6 months miniscript developers might tweak the representation as they don't expect it to be *consensus* stable, and we will be kind of stuck.

File: fedimint-api/src/encoding/tbs.rs
Comment (dpc) [inline fedimint-api/src/encoding/tbs.rs:72]: :heart: 

Could we have `Decodeable` for these as well? We might get rid of `SerdeEncodable` in some places if we did.

**Comment by dpc** [comment]
Looking great now. Pointed out one `SerdeEncodable`. The conflict should make your life easier as some of your changes won't be required anymore.

File: fedimint-api/src/encoding/btc.rs
Comment (jkitman) [reply fedimint-api/src/encoding/btc.rs:43]: I should be able to do this since we only use one kind of multisig descriptor.

File: fedimint-api/src/encoding/tbs.rs
Comment (jkitman) [reply fedimint-api/src/encoding/tbs.rs:71]: I'd save that for another PR, but yeah would be a good sanity check as well to see if they can be decoded.

File: fedimint-api/src/encoding/btc.rs
Comment (dpc) [reply fedimint-api/src/encoding/btc.rs]: We might want to put it in an enum just to leave a room for future extension.

Comment (jkitman) [reply fedimint-api/src/encoding/btc.rs]: The next change will probably be to switch to taproot then we can get rid of this implementation.

**Pull request merged**
